pkg/apis/pipeline/v1alpha1/gcs_resource.go

```diff
@@ -97,57 +97,68 @@ func (s *GCSResource) Replacements() map[string]string {
 	}
 }
 
-// GetUploadSteps gets container spec for gcs resource to be uploaded like
+// GetUploadStepModifier gets container spec for gcs resource to be uploaded like
 // set environment variable from secret params and set volume mounts for those secrets
-func (s *GCSResource) GetUploadSteps(sourcePath string) ([]Step, error) {
-	var args []string
-	if s.TypeDir {
-		args = []string{"-args", fmt.Sprintf("rsync -d -r %s %s", sourcePath, s.Location)}
-	} else {
-		args = []string{"-args", fmt.Sprintf("cp %s %s", filepath.Join(sourcePath, "*"), s.Location)}
-	}
-
-	envVars, secretVolumeMount := getSecretEnvVarsAndVolumeMounts(s.Name, gcsSecretVolumeMountPath, s.Secrets)
-
-	return []Step{{Container: corev1.Container{
-		Name:         names.SimpleNameGenerator.RestrictLengthWithRandomSuffix(fmt.Sprintf("upload-%s", s.Name)),
-		Image:        *gsutilImage,
-		Command:      []string{"/ko-app/gsutil"},
-		Args:         args,
-		VolumeMounts: secretVolumeMount,
-		Env:          envVars,
-	}}}, nil
-}
+func (s *GCSResource) GetUploadStepModifier() StepModifier {
+	return func(ts TaskSpec, sourcePath string) (*TaskSpec, error) {
+		var args []string
+		if s.TypeDir {
+			args = []string{"-args", fmt.Sprintf("rsync -d -r %s %s", sourcePath, s.Location)}
+		} else {
+			args = []string{"-args", fmt.Sprintf("cp %s %s", filepath.Join(sourcePath, "*"), s.Location)}
+		}
 
-// GetDownloadSteps returns an array of container specs to download gcs storage object
-func (s *GCSResource) GetDownloadSteps(sourcePath string) ([]Step, error) {
-	if sourcePath == "" {
-		return nil, xerrors.Errorf("GCSResource: Expect Destination Directory param to be set %s", s.Name)
-	}
-	var args []string
-	if s.TypeDir {
-		args = []string{"-args", fmt.Sprintf("rsync -d -r %s %s", s.Location, sourcePath)}
-	} else {
-		args = []string{"-args", fmt.Sprintf("cp %s %s", s.Location, sourcePath)}
+		envVars, secretVolumeMount := getSecretEnvVarsAndVolumeMounts(s.Name, gcsSecretVolumeMountPath, s.Secrets)
+
+		ts.Steps = append(
+			[]Step{{Container: corev1.Container{
```

Comment:
I'm curious about the ordering here. Wouldn't an upload occur after a task has done its work?  Looks like it's happening before the other steps here?

Reply:
Nice catch! Yes!